ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -138,18 +131,63 @@
     this.notifier = notifier;
     this.scheduler = scheduler;
     this.config = config;
+    this.accountInfoMapRef = new AtomicReference<>(new AccountInfoMap(accountServiceMetrics));
+    this.backup = new LocalBackup(this.accountServiceMetrics, config);
 
-    backupDirPath = config.backupDir.isEmpty() ? null : Files.createDirectories(Paths.get(config.backupDir));
     if (notifier != null) {
       notifier.subscribe(ACCOUNT_METADATA_CHANGE_TOPIC, changeTopicListener);
     } else {
       logger.warn("Notifier is null. Account updates cannot be notified to other entities. Local account cache may not "
           + "be in sync with remote account data.");
       accountServiceMetrics.nullNotifierCount.inc();
     }
+    if (config.useNewZNodePath) {
+      accountMetadataStore = new RouterStore(this.accountServiceMetrics, backup, helixStore, router);
+      // postpone initializeFetchAndSchedule to setupRouter function.
+    } else {
+      accountMetadataStore = new LegacyMetadataStore(this.accountServiceMetrics, backup, helixStore);
+      initialFetchAndSchedule();
+    }
+  }
+
+  /**
+   * set the router to the given one. This is a blocking call. This function would block until it fetches the
+   * {@link Account} metadata from ambry server.
+   * @param router The router to set.
+   * @throws IllegalStateException when the router already set up.
+   */
+  public void setupRouter(final Router router) throws IllegalStateException {
+    if (!this.router.compareAndSet(null, router)) {
+      throw new IllegalStateException("Router already initialized");
+    } else {
+      initialFetchAndSchedule();
+    }
+  }
+
+  /**
+   * A synchronized function to fetch account metadata from {@link AccountMetadataStore} and update the in memory cache.
+   * @param isCalledFromListener True is this function is invoked in the {@link TopicListener}.
+   */
+  private synchronized void fetchAndUpdateCache(boolean isCalledFromListener) {
+    Map<String, String> accountMap = accountMetadataStore.fetchAccountMetadata();
```

Comment:
Although this is `synchronized` method, there might be race condition when `initialFetchAndSchedule` is called before `accountMetadataStore` is instantiated. My suggestion is, if possible, move `notifier.subscribe()` to the end of constructor.

Reply:
Good Idea, I move the notifier.subscribe into the initialFetchAndSchedule function so there will be no race condition here.